julia/mmtk_julia.c

```diff
@@ -297,6 +297,160 @@ static void add_node_to_tpinned_roots_buffer(RootsWorkClosure* closure, RootsWor
     }
 }
 
+#define ADD_GLOBALLY_ROOTED(roots, r) roots[i] = (void*)r; i++;
+
+// This is a list of global variables that are marked with JL_GLOBALLY_ROOTED. We need to make sure that they
+// won't be moved. Otherwise, when we access those objects from the C global variables, we may see moved references.
+void mmtk_get_globally_rooted(void** roots, int* n)
+{
+    int i = 0;
+    ADD_GLOBALLY_ROOTED(roots, cmpswap_names);
+    ADD_GLOBALLY_ROOTED(roots, jl_typeinf_func);
+    ADD_GLOBALLY_ROOTED(roots, _jl_debug_method_invalidation);
+
+    // Max 4096
+    for (int j = 0; j < N_CALL_CACHE; j++) {
+        ADD_GLOBALLY_ROOTED(roots, call_cache[j]);
+    }
+
+    // julia_internal.h
+    ADD_GLOBALLY_ROOTED(roots, jl_type_type_mt);
+    ADD_GLOBALLY_ROOTED(roots, jl_nonfunction_mt);
+    ADD_GLOBALLY_ROOTED(roots, jl_kwcall_mt);
+    ADD_GLOBALLY_ROOTED(roots, jl_all_methods);
+    ADD_GLOBALLY_ROOTED(roots, jl_module_init_order);
+    // ADD_GLOBALLY_ROOTED(roots, jl_current_modules); -- we cannot trace a htable_t. But all the modules are pinned. So we are fine.
+    ADD_GLOBALLY_ROOTED(roots, jl_precompile_toplevel_module);
+    ADD_GLOBALLY_ROOTED(roots, jl_global_roots_table);
```

Comment:
For the global roots table, do we need to go through the table and add each object in it? It looks like the code below is still transitively pinning this, so I'm not sure you'd want to fix it in this PR.

Reply:
We still transitive pin global roots table. So we don't have to pin each single item in the table. 

This extra step is to ensure we treat all the 'JL_GLOBALLY_ROOTED' symbol in the same way. I think they have to be pinned, otherwise the C global would reference an invalid object.

Reply:
That brings another question, do you actually need to add those roots separately in the `scan_vm_specific_roots` function? Are they subsumed by the global roots that are tagged by `JL_GLOBALLY_ROOTED`, i.e., should we then replace the code in that function by the code you added instead? 
Also, I see [here](https://github.com/mmtk/julia/blob/f5f8510c016b40cb82f7445338bf9ef084704174/src/julia.h#L1798) that the `jl_main_module` is tagged with `JL_GLOBALLY_ROOTED` but it doesn't seem to be on your list. Are you purposefully skipping modules because they'll be pinned? (If you are, then we can't replace the code in `scan_vm_specific_roots` since `jl_main_module` is added there and not by `ADD_GLOBALLY_ROOTED`)

Reply:
> I see, but some of these overlap with the roots added in `scan_vm_specific_roots` such as `cmpswap_names`, `jl_an_empty_vec_any`, `jl_module_init_order`, etc. I'm not sure if it's a big deal to add them twice, but maybe leave out or simply comment out the ones that are already added in that function?

I see. Those code are copied from Julia's GC -- they only trace some of the globally rooted symbols, and assume the rest will be reachable from the existing roots. But that is not sufficient for a moving GC.

Reply:
> Also, I see [here](https://github.com/mmtk/julia/blob/f5f8510c016b40cb82f7445338bf9ef084704174/src/julia.h#L1798) that the `jl_main_module` is tagged with `JL_GLOBALLY_ROOTED` but it doesn't seem to be on your list. Are you purposefully skipping modules because they'll be pinned? (If you are, then we can't replace the code in `scan_vm_specific_roots` since `jl_main_module` is added there and not by `ADD_GLOBALLY_ROOTED`)

Umm, I somehow missed those. I will check again and make sure all the symbols with `JL_GLOBALLY_ROOTED` are traced.

Reply:
> > I see, but some of these overlap with the roots added in `scan_vm_specific_roots` such as `cmpswap_names`, `jl_an_empty_vec_any`, `jl_module_init_order`, etc. I'm not sure if it's a big deal to add them twice, but maybe leave out or simply comment out the ones that are already added in that function?
> 
> I see. Those code are copied from Julia's GC -- they only trace some of the globally rooted symbols, and assume the rest will be reachable from the existing roots. But that is not sufficient for a moving GC.

Yeah, then I'd say we should replace the code in `scan_vm_specific_roots` by just a call to `mmtk_get_globally_rooted`, and you'll possibly need to add the following code in there to process the rooted modules:
```
for (size_t j = 0; j < jl_current_modules.size; j += 2) {
    if (jl_current_modules.table[j + 1] != HT_NOTFOUND) {
        ADD_GLOBALLY_ROOTED(roots, jl_current_modules.table[j]);
    }
}
```
You can add a comment to say that the roots in `mmtk_get_globally_rooted` are a superset of the roots from [here](https://github.com/mmtk/julia/blob/f5f8510c016b40cb82f7445338bf9ef084704174/src/gc.c#L2613), that must be passed to the GC when we're moving objects since these references need to be updated.